### DISCUSSION TITLE
BLE Service Hierarchy

### DISCUSSION THREAD

**Discussion by AGlass0fMilk** [ARMmbed/mbed-os-experimental-ble-services#40] [Ideas]
Something I think should be discussed before we make a large library of BLE services is their class hierarchy.

Should all services share the same base class? Is there a use case where this would be advantageous?

I would also like to discuss how services can be extended.

For example, let's take the WIP DFU Service (see #13). Our goal is to provide a basic DFU implementation that works on all BLE-enabled targets. However, there are some cases where the performance of the DFU process can be improved by introducing some extension. An example of an extension in this case:

@pan- [suggested](https://github.com/ARMmbed/mbed-os-experimental-ble-services/pull/13#issuecomment-733109997) a control point that lets the GATT client request the GATT server to perform connection parameter updates.

> A service that allows a client to make the server send a connection parameter update request would be a useful addition, mobile platforms generally accepts new connection parameters but do not have options to tune it.

Additionally, as it stands now, the DFU service implementation has a delta mode feature that not all targets may implement. Therefore, I think features like this should be considered extensions of the base DFU service.

I would like to come up with a recommended pattern of creating reusable and extensible service classes.

Taking the DFUService as an example again: 

We need some way for the child class to add characteristics to the service's array that is submitted to the BLE stack.

How can we do this? My proposed solution would be to have a `virtual` method that child classes could override. This would return a `Span` of `GattCharacteristic`s to add to the array. The array could be dynamically allocated or have an upper limit on maximum number of characteristics set statically in a configuration. The array of characteristics is copied internally by the stack and so it can be discarded after use.

@pan- @paul-szczepanek-arm @noonfom Do any of you have comments/ideas for what the hierarchy of BLE service classes should be?

**Comment by pan-** [comment]
It rejoins the Bluetooth framework discussion. All Bluetooth services will require access to the same basic services: 
- EventQueue.
- Ble instances (GattServer, ...).
- Chain-able event handlers: Gap, GattServer.

Services will follow a common lifecycle too: 
- Initialization after BLE has been initialized.
- Termination when the system is about to shutdown BLE.

We probably want an entity that aggregates the service, contain their common dependencies and drive their state (initialization/termination). A base service class can be associated with it:

```c++
struct ServiceController;
struct Service;

struct ServiceController { 
    void addService(Service& service)  {
        Service& *it = _services;
        while (it) { 
            it = it->next;
        }
        it = service;
       service.setController(this);
   } 
 
    void removeService(Service&);

     void init() { 
        Service& *it = _services;
        while (it) { 
            it = it->init();
        }
      }

protected:
    friend class Service;

   // Access to base dependency
   GattServer& server();
   EventQueue& queue();

private:
    Service* _services;
    ChainableGattEventHandler* _handlersChain;  
};


struct Service : GattServer::EventHandler { 
     / /Constructor


protected: 
     friend class ServiceController;

     // Called by the controller when services are ready to be registered
     virtual void init() = 0;

     // Called by the controller when Bluetooth is being shutdown
     virtual void shutdown() = 0;

     void setController(ServiceController* controller) { 
          _controller = controller;
     }

    // Access to the dependencies through the controller
   GattServer& server();
   EventQueue& queue();

private:
     ServiceController* _controller = nullptr;
     Service* _next = nullptr; // Next service. Used by the controller 
};
```

For the second part of your question, a base class can probably offer functions to return the base `GattService` and `GattCharacteristics` while the derived class assemble these pieces when it is initialized. Structure of the characteristic is known so we probably don't need to dynamically allocate them: 

```c++
struct FooService { 
     virtual void init() { 
          auto service = makeService();
          auto characteristics = makeCharacteristics(); 
          
          // Assemble base characteristics and service
     }

protected:
    GattService makeService();

    FooCharacteristic makeCharacteristics() {
        // initialize and return the characteristics
    }

    struct FooCharacteristics { 
        GattCharacteristic baz;
        GattCharacteristic qux;
    };
};

struct BarService : FooService { 
     virtual void init() { 
          auto service = makeService();    // From base class 
          auto characteristics = makeCharacteristics(); // From derived class 
          
          // Assemble base characteristics and service
     }

protected:
    BarCharacteristics makeCharacteristics() {
        // initialize and return the characteristics
       // Call Foo::makeCharateristic to get the base characteristics
    }

    struct BarCharacteristics { 
        struct FooCharacteristics fooCharacteristics;
        GattCharacteristic waldo; 
    };
};
```

I don't expect in service extension to be very common, it is probably easier to add flags to the service and manage them inside the characteristic. It can become complex to override behavior based on events received... And it would makes client harder to develop too.
